openmetrics/src/main/java/io/airlift/openmetrics/MetricsResource.java

```diff
@@ -0,0 +1,281 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.openmetrics;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import io.airlift.openmetrics.types.BigCounter;
+import io.airlift.openmetrics.types.Counter;
+import io.airlift.openmetrics.types.Gauge;
+import io.airlift.openmetrics.types.Info;
+import io.airlift.openmetrics.types.Metric;
+
+import javax.inject.Inject;
+import javax.management.InstanceNotFoundException;
+import javax.management.IntrospectionException;
+import javax.management.JMException;
+import javax.management.MBeanAttributeInfo;
+import javax.management.MBeanInfo;
+import javax.management.MBeanServer;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectName;
+import javax.management.ReflectionException;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+@Path("/v1/metrics")
+public class MetricsResource
+{
+    private static final Logger log = Logger.get(MetricsResource.class);
+    private static final String OPENMETRICS_CONTENT_TYPE = "application/openmetrics-text; version=1.0.0; charset=utf-8";
+    private static final String ATTRIBUTE_SEPARATOR = "_ATTRIBUTE_";
+    private static final String TYPE_SEPARATOR = "_TYPE_";
+    private static final String NAME_SEPARATOR = "_NAME_";
+    private static final Pattern METRIC_NAME_PATTERN = Pattern.compile("[[a-zA-Z]][\\w_]*");
+    private static final LoadingCache<String, Boolean> VALID_NAMES_CACHE = CacheBuilder.newBuilder().build(new CacheLoader<String, Boolean>()
+    {
+        @Override
+        public Boolean load(String metricName)
+        {
+            return METRIC_NAME_PATTERN.matcher(metricName).matches();
+        }
+    });
+
+    private final MBeanServer mbeanServer;
+    private final boolean inferNumeric;
+    private final boolean inferString;
+    private final boolean inferAll;
+
+    private final List<ObjectName> allMetricsObjectNames;
+
+    @Inject
+    public MetricsResource(MBeanServer mbeanServer, MetricsConfig metricsConfig)
+    {
+        this.mbeanServer = requireNonNull(mbeanServer, "mbeanServer is null");
+        this.inferNumeric = metricsConfig.isInferNumeric();
+        this.inferString = metricsConfig.isInferStrings();
+        this.inferAll = metricsConfig.isInferAll();
```

Comment:
When I read `@ConfigDescription` i though that `inferAll` implies `inferNumeric` and `inferString`.
If that is not the case maybe rename the property to `inferOther`.

Or change above lines to:
```
        this.inferNumeric = metricsConfig.isInferAll() || metricsConfig.isInferNumeric();
        this.inferString = metricsConfig.isInferAll() || metricsConfig.isInferStrings();
```